linkerd/stack/src/queue.rs

```diff
@@ -0,0 +1,117 @@
+use crate::{
+    failfast::{self, FailFast},
+    layer::{self, Layer},
+    BoxService, CloneParam, ExtractParam, NewService, Param, Service,
+};
+use linkerd_error::Error;
+use std::{fmt, marker::PhantomData, time::Duration};
+use tower::buffer::Buffer;
+
+#[derive(Debug, Copy, Clone)]
+pub struct QueueConfig {
+    /// The number of requests (or connections, depending on the context) that
+    /// may be buffered
+    pub capacity: usize,
+
+    /// The maximum amount of time a request may be buffered before failfast
+    /// errors are emitted.
+    pub failfast_timeout: Duration,
+}
+
+pub struct NewQueue<N, Req, X = ()> {
+    inner: N,
+    extract: X,
+    _req: PhantomData<fn(Req)>,
+}
+
+pub type Queue<Req, Rsp> = failfast::Gate<Buffer<BoxService<Req, Rsp, Error>, Req>>;
+
+// === impl NewQueue ===
+
+impl<T, N, X, Req> NewService<T> for NewQueue<N, Req, X>
+where
+    Req: Send + 'static,
+    X: ExtractParam<QueueConfig, T>,
+    N: NewService<T>,
+    N::Service: Service<Req> + Send + 'static,
+    <N::Service as Service<Req>>::Future: Send,
+    <N::Service as Service<Req>>::Error: Into<Error>,
+    <N::Service as Service<Req>>::Response: 'static,
+{
+    type Service = Queue<Req, <N::Service as Service<Req>>::Response>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let QueueConfig {
+            capacity,
+            failfast_timeout,
+        } = self.extract.extract_param(&target);
+        let buf = layer::mk(move |inner| Buffer::new(BoxService::new(inner), capacity));
+        let buf = FailFast::layer_gated(failfast_timeout, buf);
+        buf.layer(self.inner.new_service(target))
+    }
+}
+
+impl<T: Param<QueueConfig>, Req> NewQueue<T, Req> {
```

Comment:
This type signature doesn't look right to me. `T` here is the NewService type. We don't want any type constraints here. In general, we should use `N` for the primary inner new service.